### PR TITLE
refactor

### PR DESCRIPTION
**Pull request by nodatall**
> In order to succeed, we must first believe that we can refactor
~ Nikos Kazantzakis

### CODE REVIEW

File: src/factorial.js [hunk 1]
```diff
@@ -4,8 +4,16 @@ const factorial = number => {
...
   return result
 }
 
-export default factorial
+const recursiveFactorial = number => {
+  if ( number === 1 ) {
+    return number
+  }
+
+  return recursiveFactorial( number - 1 ) * number
```
Comment (jrobcodes) [inline src/factorial.js:16]: This is sort of a trivial thing, but this reads better to me if written as:
```js
return number * recursiveFactorial( number - 1 )
```
This also [allows for optimizations](https://en.wikipedia.org/wiki/Tail_call) (although the compiler is probably smart enough to do this by itself).

File: src/fizzBuzz.js [hunk 2]
```diff
@@ -1,16 +1,19 @@
...
+    stringyString = ''
+
+    if ( index % 3 === 0) {
+      stringyString += 'Fizz'
+    }
+    if ( index % 5 === 0 ) {
+      stringyString += 'Buzz'
     }
+    
+    stringyString.length ? result.push(stringyString) : result.push(index)
```
Comment (jrobcodes) [inline src/fizzBuzz.js:16]: I prefer explicit `if`/`else` blocks over ternary operations, especially in this case where we are performing operations instead of a calculation for eventual assignment.  As well, I prefer to be explicit about the boolean expression, instead of relying on the truthiness/falsiness of a value (is zero false?).

File: src/isPalindrome.js [hunk 3]
```diff
@@ -1,14 +1,28 @@
-const isPalidrome = string => {
+const isPalindrome = ( string, callbackOne, callbackTwo, callbackThree ) => {
```
Comment (jrobcodes) [inline src/isPalindrome.js:1]: This implementation is confusing to me, and does not seem to meet the specification provided in the goal?  It seems to me as if we are forced, as a user of the `isPalindrome` function, to understand *how* `isPalindrome` works, and to call it by passing in these additional callback functions...  Couldn't we just omit these from the parameters, and call them directly, instead of exporting them and forcing the user to inject them?

File: src/makeChange.js [hunk 4]
```diff
@@ -1,32 +1,52 @@
-const makeChange = ({price, amountGiven}) => {
-  let change = { quarters: 0, dimes: 0, nickels: 0, pennies: 0 },
-      remainder = amountGiven - price
-
-  const findChangeAmount = ( denomination, amount ) => {
-    let numberOfDenomination = Math.floor( remainder / amount )
-    change[denomination] = numberOfDenomination
-    remainder = remainder - ( numberOfDenomination * amount )
+const makeChange = ({ price, amountGiven, callback }) => {
```
Comment (jrobcodes) [inline src/makeChange.js:1]: Here's another case where we are forcing users of this function to explicitly understand how this function works in order to be able to use it... Consider removing the injected callback, and invoking the function explicitly as part of the implementation.

File: test/factorial_test.js [hunk 5]
```diff
@@ -1,5 +1,5 @@
 import { expect } from 'chai'
-import factorial from '../src/factorial'
+import { factorial, recursiveFactorial } from '../src/factorial'
```
Comment (jrobcodes) [inline test/factorial_test.js:2]: We should also test negative cases like `factorial( 0 )` and `factorial( -1 )` to ensure this function works when called with invalid parameters.
